lib/pbench/agent/utils.py

```diff
@@ -396,6 +398,58 @@ def collect_local_info(pbench_bin):
     return (version, seqno, sha1, hostdata)
 
 
+class TemplateSsh:
+    """
+    Set up to easily launch repeated asynchronous ssh commands from a template
+    and acquire the stdout and stderr streams along with the completion status.
+    """
+
+    def __init__(self, ssh_cmd: Path, ssh_args: List[str], cmd: List[str]):
+        """
+        Create an SSH template object
+
+        Args:
+            ssh_cmd: Path to the ssh command
+            ssh_args: A partial argv representing ssh command options
+            cmd: A partial argv representing a remote command to be executed
+        """
+        self.command = cmd
+        self.procs: Dict[str, subprocess.Popen] = {}
+        self.base_args = [ssh_cmd] + ssh_args
+
+    def start(self, host: str):
+        """
+        Begin an asynchronous ssh command on the specified remote host
+
+        Args:
+            host: hostname or IP
+        """
+        args = self.base_args + [host] + self.command
```

Comment:
Should we be [quoting](https://docs.python.org/3/library/shlex.html#shlex.quote) (or joining) `self.command`?  (Or, accepting it as a single string, instead of a list, and expecting the caller to do it?)

Reply:
I looked at what Peter did and ultimately decided that I liked yours better (because I couldn't come up with a really good way to "remember" which element was the hostname...the best I could come up with was to use a `dict` instead of a `list`, which the code could access by key-name and then reference the `values()` of it for the command invocation, but what you've got seemed better (although, now that I think of it, a `namedtuple` would work well, since you can access the components by name but also iterate over the components like a list, so you might be able to pass it directly to `Popen()` (or worst-case, apply the `list()` constructor to it))).

But, what I'm getting at with my comment here is that `base_args` + `host` are the arguments to the `ssh` command, and it takes **_one_** more argument, which is the remote command to be run.  It shouldn't take a list of components for that command, because (I think) we don't want to feed additional components to the `ssh` command.  (I was reading about `shlex.split()` which led to `shlex.quote()` whose code-injection example led me to ponder your command argument to `ssh`....)

Reply:
> `ssh root@host ls /tmp` and `ssh root@host 'ls /tmp'` are exactly the same.

So it is guaranteed that no substitutions are made? How? Do we have to examine the callers?

Reply:
Substitution depends on whether you're running the command through a shell or "bare", as well as on the tokens you use (regardless of argv breakdown vs string) are quoted. I'm not running the `subprocess.Popen` with `shell=True` so there's no local shell in any case.

There's a single caller for the new class (although I've toyed with the idea of converting the `pbench-tool-meister-start` method `start_tms_via_ssh` because I think this framework would be cleaner ... if I expand the templating mechanism to allow altering the command as well as the host). The single caller just wants to pass `echo ${SSH_CONNECTION}` to be expanded on the remote so we get the SSH origin hostname, which provides our guarantee of external routability.

Reply:
> I'm not running the subprocess.Popen with shell=True so there's no local shell in any case.

That's a "feature".  (That also allays most of my concern...and, it also means that there's no confusion about where `${SSH_CONNECTION}` will be evaluated.)